In [1]:
import sys
sys.path.append("../") # go to parent dir
from util.read_data import DataReader
from util.evaluator import ModelEvaluator


In [2]:
reader = DataReader()
df = reader.read_all_data_normalized()
evaluator = ModelEvaluator(reader.label_scaler)
df.columns

Index(['Open', 'High', 'Low', 'Close', 'Volume', 'DayOfWeek', 'Month',
       'AdjClose', 'OpenDiff', 'CloseDiff', 'AdjCloseDiff', 'HighLowDiff'],
      dtype='object')

In [3]:
df_test_features, series_test_volume = reader.get_test_data(df)


Baseline model: $\hat{E}[v_{d+1}]=v_d$

In [4]:

def baseline_model(volume):
    pred_volume = volume.shift(1)
    return pred_volume


In [5]:
baseline_pred_volume = baseline_model(series_test_volume)
baseline_model_results = evaluator.evaluate("baseline", series_test_volume[1:], baseline_pred_volume[1:])
print(baseline_model_results)

baseline: MSE = 3.684400e+17, R2 = 0.033, confidence interval 95% = (-53,069,427 - 53,341,067)


The baseline r2 score is **0.033**

Another simple model would use moving average over some time window.
The model is following: $V_{d+1}=\frac 1n \sum_{i=0}^{n-1} V_{d-i}$.
We can try multiple n values, le'ts try following values: {1, 3, 5, 10, 50, 100}



In [6]:
def moving_average_model(volume, n):
    pred_volume = volume.rolling(n).mean().shift(1)
    return pred_volume

In [7]:
window_size = [1,3,5,10,50,100]
moving_average_result_list = [evaluator.evaluate("moving_average {}".format(n), 
                                                 moving_average_model(series_test_volume, n)[n:], series_test_volume[n:]) 
                                                 for n in window_size]

for result in moving_average_result_list:
    print(result)


moving_average 1: MSE = 3.684400e+17, R2 = 0.033, confidence interval 95% = (-53,341,067 - 53,069,427)
moving_average 3: MSE = 3.099159e+17, R2 = -0.239, confidence interval 95% = (-48,259,465 - 49,530,282)
moving_average 5: MSE = 3.128615e+17, R2 = -0.523, confidence interval 95% = (-45,836,522 - 52,614,748)
moving_average 10: MSE = 3.298599e+17, R2 = -1.303, confidence interval 95% = (-39,324,762 - 62,279,075)
moving_average 50: MSE = 3.892483e+17, R2 = -6.839, confidence interval 95% = (-23,387,014 - 91,775,513)
moving_average 100: MSE = 4.390438e+17, R2 = -23.992, confidence interval 95% = (-11,765,661 - 117,942,651)


The moving average models for window size <3, 50> does not help much. The window size 1 at least confirms that te algorithm has correct implementation.

Le'ts try to use exponential moving average, 

In [8]:
def exp_moving_average_model(volume, n):
    pred_volume = volume.ewm(span=n).mean().shift(1)
    return pred_volume



In [9]:
window_size = [1,3,5,10,50,100]
exp_moving_average_result_list = [evaluator.evaluate("moving_average {}".format(n),
                                                 exp_moving_average_model(series_test_volume, n)[n:], series_test_volume[n:])
                                                 for n in window_size]

for result in exp_moving_average_result_list:
    print(result)


moving_average 1: MSE = 3.684400e+17, R2 = 0.033, confidence interval 95% = (-53,341,067 - 53,069,427)
moving_average 3: MSE = 2.880625e+17, R2 = -0.213, confidence interval 95% = (-46,252,592 - 48,026,359)
moving_average 5: MSE = 2.845396e+17, R2 = -0.489, confidence interval 95% = (-42,536,855 - 51,352,553)
moving_average 10: MSE = 2.953661e+17, R2 = -1.156, confidence interval 95% = (-38,026,698 - 58,118,071)
moving_average 50: MSE = 3.594106e+17, R2 = -6.870, confidence interval 95% = (-25,767,251 - 84,893,406)
moving_average 100: MSE = 4.073537e+17, R2 = -18.608, confidence interval 95% = (-20,472,852 - 104,466,639)


Exponential moving average on the Volume does not help much either.

Let's not try to use first real machine-learning model, random forest classifier. The features will be month of year, day of week, difference between low and high price of previous day and historical parametrized window of following market data: volume data, and changes in the adjusted closing price. Size of the window will be: {3, 5, 10, 50, 100}. The expected result is that the accuracy will be better with longer time window, but with deminishing returns.

In [10]:
from sklearn.linear_model import LinearRegression

def linear_regression_model(features, volume):
    regressor = LinearRegression()
    
    regressor.fit(features, volume)
    return regressor

def eval_model(name, model_function, n, custom_callback=None):
    # prepare the features
    df_enriched = reader.prepare_window_features_for_training(df, n)[n:]

    train_features, train_volume = reader.get_train_data(df_enriched)
    test_features, test_volume = reader.get_test_data(df_enriched)

    regressor = model_function(train_features, train_volume)
    
    if (custom_callback):
        random_forest_callback(train_features, regressor)
   
    print(evaluator.evaluate("{} {} on train: ".format(name, n), regressor.predict(train_features), train_volume))
    result = evaluator.evaluate("{} {} on test:".format(name, n), regressor.predict(test_features), test_volume)
    print(result)
    print()
    return result


In [11]:
for n in (1,3,5,10,50,100):
    eval_model("linear_regression", linear_regression_model, n)

linear_regression 1 on train: : MSE = 3.366615e+17, R2 = 0.841, confidence interval 95% = (-17,391,377 - 17,391,377)
linear_regression 1 on test:: MSE = 3.159721e+17, R2 = 0.037, confidence interval 95% = (-11,065,462 - 87,379,065)

linear_regression 3 on train: : MSE = 2.925643e+17, R2 = 0.865, confidence interval 95% = (-16,216,214 - 16,216,214)
linear_regression 3 on test:: MSE = 2.649089e+17, R2 = 0.008, confidence interval 95% = (-20,038,420 - 70,101,150)

linear_regression 5 on train: : MSE = 2.829278e+17, R2 = 0.870, confidence interval 95% = (-15,944,029 - 15,957,264)
linear_regression 5 on test:: MSE = 2.594226e+17, R2 = -0.037, confidence interval 95% = (-22,011,386 - 67,189,891)

linear_regression 10 on train: : MSE = 2.737411e+17, R2 = 0.874, confidence interval 95% = (-15,713,567 - 15,683,907)
linear_regression 10 on test:: MSE = 2.498021e+17, R2 = -0.119, confidence interval 95% = (-23,543,374 - 63,988,302)

linear_regression 50 on train: : MSE = 2.587606e+17, R2 = 0.881,

Linera regression does not seem very good results, with 

In [12]:
from sklearn.ensemble import RandomForestRegressor

def random_forest_model(features, volume):
    regressor = RandomForestRegressor(n_estimators=100)
    
    regressor.fit(features, volume)
    return regressor

def random_forest_callback(train_features, regressor):
    importances = list(zip(train_features.columns, regressor.feature_importances_))
    importances.sort(key=lambda x: x[1], reverse = True)
    print("First 5 Feature importances {}:".format(n))
    for importance in importances[:5]:
        print('Feature: {:30} \t Importance: {}'.format(importance[0], importance[1]))


In [13]:
for n in (1,3,5,10,20):
    eval_model("random_forest", random_forest_model, n, random_forest_callback)

First 5 Feature importances 1:
Feature: Volume1                        	 Importance: 0.9048789127636978
Feature: AdjCloseDiff1                  	 Importance: 0.04084382139957442
Feature: DayOfWeek_Monday               	 Importance: 0.009608795418870848
Feature: Month_12                       	 Importance: 0.005036360290943819
Feature: DayOfWeek_Tuesday              	 Importance: 0.004456825398871814
random_forest 1 on train: : MSE = 4.929654e+16, R2 = 0.978, confidence interval 95% = (-9,753,549 - 3,556,371)
random_forest 1 on test:: MSE = 2.841516e+17, R2 = 0.065, confidence interval 95% = (-112,584,593 - -19,228,597)

First 5 Feature importances 3:
Feature: Volume1                        	 Importance: 0.8314604159589224
Feature: Volume2                        	 Importance: 0.0662501170767072
Feature: Volume3                        	 Importance: 0.030515485600666706
Feature: AdjCloseDiff1                  	 Importance: 0.016658033099810426
Feature: AdjCloseDiff3                  	 Imp

It is obvious from the comparison between train and test results that the model has high variance, it overfits to the train set. The decision tree regression model is not suitable for the task.  is Le't see SVM model, for comparison.

In [14]:
from sklearn.svm import SVR

In [15]:
def svm_model(features, volume):
    regressor = SVR()
    
    regressor.fit(features, volume)
    return regressor


In [16]:
for n in (1, 2, 3,5,10,20):
    eval_model("svm_model", svm_model, n)

/home/martin/work/github/trade_volume_prediction/venv/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


svm_model 1 on train: : MSE = 2.987715e+17, R2 = 0.857, confidence interval 95% = (22,203,979 - 54,970,986)
svm_model 1 on test:: MSE = 2.841244e+17, R2 = 0.135, confidence interval 95% = (-27,130,374 - 66,221,161)



/home/martin/work/github/trade_volume_prediction/venv/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


svm_model 2 on train: : MSE = 2.579920e+17, R2 = 0.877, confidence interval 95% = (18,432,532 - 48,884,906)
svm_model 2 on test:: MSE = 2.460071e+17, R2 = 0.107, confidence interval 95% = (-15,977,399 - 70,886,847)



/home/martin/work/github/trade_volume_prediction/venv/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


svm_model 3 on train: : MSE = 2.479718e+17, R2 = 0.882, confidence interval 95% = (14,772,783 - 44,631,416)
svm_model 3 on test:: MSE = 2.466826e+17, R2 = 0.048, confidence interval 95% = (-14,763,232 - 72,220,184)



/home/martin/work/github/trade_volume_prediction/venv/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


svm_model 5 on train: : MSE = 2.272681e+17, R2 = 0.892, confidence interval 95% = (15,414,757 - 44,006,437)
svm_model 5 on test:: MSE = 2.611143e+17, R2 = 0.006, confidence interval 95% = (-32,201,483 - 57,290,172)



/home/martin/work/github/trade_volume_prediction/venv/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


svm_model 10 on train: : MSE = 2.040400e+17, R2 = 0.903, confidence interval 95% = (12,209,797 - 39,316,855)
svm_model 10 on test:: MSE = 2.597789e+17, R2 = -0.081, confidence interval 95% = (-47,058,810 - 42,203,708)



/home/martin/work/github/trade_volume_prediction/venv/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


svm_model 20 on train: : MSE = 1.803236e+17, R2 = 0.913, confidence interval 95% = (12,556,266 - 38,069,206)
svm_model 20 on test:: MSE = 2.738548e+17, R2 = -0.495, confidence interval 95% = (-42,276,160 - 49,372,772)



The best SVM model is again the least overfitted, with only latest value of Volume and AdjCloseDiff1, and seasonality features (DayOfWeek, Month). The R2 = 0.135, with confidence interval 95% of average error = (-27,130,374 - 66,221,161)

It seem that we are not moving forward a lot with traditional ML models.
Let's try one more model, improved ridge regression with cross validation.

In [17]:
from sklearn import linear_model

def linear_ridge_model(features, volume):
    regressor = linear_model.RidgeCV(alphas=[0.1, 1.0, 10.0], cv=10)
    
    regressor.fit(features, volume)
    return regressor

In [18]:
for n in (1, 2, 3,5,10,20):
    eval_model("linear_ridge_model", linear_ridge_model, n)

linear_ridge_model 1 on train: : MSE = 3.366615e+17, R2 = 0.841, confidence interval 95% = (-17,391,377 - 17,391,377)
linear_ridge_model 1 on test:: MSE = 3.159678e+17, R2 = 0.037, confidence interval 95% = (-11,055,002 - 87,388,853)

linear_ridge_model 2 on train: : MSE = 2.989060e+17, R2 = 0.861, confidence interval 95% = (-16,389,109 - 16,389,109)
linear_ridge_model 2 on test:: MSE = 2.725049e+17, R2 = 0.034, confidence interval 95% = (-17,928,323 - 73,494,442)

linear_ridge_model 3 on train: : MSE = 2.925806e+17, R2 = 0.864, confidence interval 95% = (-16,216,666 - 16,216,666)
linear_ridge_model 3 on test:: MSE = 2.647186e+17, R2 = 0.004, confidence interval 95% = (-19,594,697 - 70,512,494)

linear_ridge_model 5 on train: : MSE = 2.829484e+17, R2 = 0.869, confidence interval 95% = (-15,951,225 - 15,951,225)
linear_ridge_model 5 on test:: MSE = 2.592607e+17, R2 = -0.041, confidence interval 95% = (-21,673,298 - 67,500,148)

linear_ridge_model 10 on train: : MSE = 2.737625e+17, R2 = 